metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -128,9 +128,9 @@ func NewResourceMetadataEnricher(
 
 	cfg, _ := common.NewConfigFrom(&metaConfig)
 
-	metaGen := metadata.NewResourceMetadataGenerator(cfg)
+	metaGen := metadata.NewResourceMetadataGenerator(cfg, watcher.Client())
 	podMetaGen := metadata.NewPodMetadataGenerator(cfg, nil, watcher.Client(), nil, nil)
-	serviceMetaGen := metadata.NewServiceMetadataGenerator(cfg, nil, nil)
+	serviceMetaGen := metadata.NewServiceMetadataGenerator(cfg, nil, nil, watcher.Client())
```

Comment:
To my mind `Generate` should have the capability to talk to the k8s API since it is already taking this direction with other metadata like ,`service.selectors` or `kubernetes.deployment` when they are not directly available in Resource's struct.

Having to fetch additional data prior to `Generate`'s call and passing them as `FieldOption` does look quite error prone and I can predict cases in the future where we will not follow the same paths from the various places we add k8s metadata. In this, I'm thinking of the generator as a single source of truth capable to add the complete set of metadata ...to be continued at https://github.com/elastic/beats/pull/26056#discussion_r647306344 :)